apps/prairielearn/src/lib/bot-grading.ts

```diff
@@ -0,0 +1,165 @@
+import { type Request, type Response } from 'express';
+import { OpenAI } from 'openai';
+
+import { logger } from '@prairielearn/logger';
+import { loadSqlEquiv, queryRow, queryRows } from '@prairielearn/postgres';
+
+import { InstanceQuestionSchema, SubmissionSchema, VariantSchema } from '../lib/db-types.js';
+import * as manualGrading from '../lib/manualGrading.js';
+import { buildQuestionUrls } from '../lib/question-render.js';
+import { getQuestionCourse } from '../lib/question-variant.js';
+import { createServerJob } from '../lib/server-jobs.js';
+import * as questionServers from '../question-servers/index.js';
+
+const sql = loadSqlEquiv(import.meta.url);
+
+export async function botGrade(
+  req: Request,
+  res: Response,
+  openaiconfig: Record<'apiKey' | 'organization', string>,
+): Promise<string> {
+  const question = res.locals.question;
+  const question_course = await getQuestionCourse(question, res.locals.course);
+
+  const serverJob = await createServerJob({
+    courseId: res.locals.course ? res.locals.course.id : null,
+    type: 'botGrading',
```

Comment:
Probably a good idea to also include the course instance ID, assessment ID, authn user ID as arguments, since this allows you to later on filter jobs by those fields.